### DISCUSSION TITLE
How can I use C# code to generate different output format?

### DISCUSSION THREAD

**Discussion by Oliverfk** [danielpalme/ReportGenerator#573] [Q&A]
I am using the 5.1.12 version on .Net 6.0. I have a JaCoCo report as an input file, and I was trying to figure out a way to generate a Cobertura format using C# code directly, which will be hosted in the server. 

I noticed there's a `CoberturaReportBuilder()` class, so I tried below code snippet:
```
  var coberturaReportBuilder = new CoberturaReportBuilder();
  var summaryResult = new SummaryResult(parserResult);
  coberturaReportBuilder.CreateSummaryReport(summaryResult);
```

However, it throws error: `_System.NullReferenceException: Object reference not set to an instance of an object.
   at Palmmedia.ReportGenerator.Core.Reporting.Builders.CoberturaReportBuilder.CreateSummaryReport(SummaryResult summaryResult) in /_/src/ReportGenerator.Core/Reporting/Builders/CoberturaReportBuilder.cs:line 231_` Maybe that is due to the `ReportContext` property is null, which cannot be initialized.

I am wondering what is the correct way to do this. Thank you.

**Comment by danielpalme** [comment]
I would recommend to use _ReportGenerator_ as a command line tool:
`ReportGenerator.exe -reports:JaCoCo.xml -targetdir:report -reporttypes:Cobertura`

If you really want to do this in code, you could use:

```csharp
string[] sourceDirectories = new string[]
{
};

string path = "JaCoCo0.8.3.xml";

var configuration = new ReportConfiguration(
    new[] { path },
    "C:\\temp",
    sourceDirectories,
    null,
    new[] { "Cobertura" },
    Array.Empty<string>(),
    Array.Empty<string>(),
    Array.Empty<string>(),
    Array.Empty<string>(),
    null,
    null);

new Generator().GenerateReport(configuration);
```

**Reply by Oliverfk** [reply]
Unfortunately, I had to do in C# in our backend hosted on Azure Function.

Questions though. The first parameter of `ReportConfiguration()` is called `reportFilePatterns`, yet the third parameter is called `sourceDirectoies`. I am a little confused why you put `sourceDirectories` as an empty string array, but have the `reportFilePatterns` to `new [] {"JaCoCo0.8.3.xml"} `? Where should I put my source files actually?

**Comment by danielpalme** [comment]
`reportFilePatterns` contains your coverage reports (here your JaCoCo file: "JaCoCo0.8.3.xml").
`sourceDirectories` is typically empty, but some coverage formats require that your supply the base directories of your **source code**.
For JaCoCo you should supply them.

**Reply by Oliverfk** [reply]
I just tested using the `SampleReports\Clover\Clover.xml` as the `reportFilePatterns` provided within the SampleReports folder, and tried to convert to Cobertura format:
```
            string[] sourceDirectories = new string[] {};

            string pattern = @"C:\Users\Downloads\SampleReports\Clover\Clover.xml";

            var configuration = new ReportConfiguration(
                new[] { pattern },
               "C:\\temp",
                sourceDirectories,
                null,
                new[] { "Cobertura" },
                Array.Empty<string>(),
                Array.Empty<string>(),
                Array.Empty<string>(),
                Array.Empty<string>(),
                null,
                null);

            new Generator().GenerateReport(configuration);
```
However, the converted result does not contain the covered lines, and it seems missing the data. I am attaching the clover.xml and the output cobertura.xml. Is this expected? Should I also have to supply the base directories of the source code?
[Clover.txt](https://github.com/danielpalme/ReportGenerator/files/10539983/Clover.txt)
[Cobertura.txt](https://github.com/danielpalme/ReportGenerator/files/10539984/Cobertura.txt)

**Reply by danielpalme** [reply]
The reason for the missing details in the _Cobertura_ file is, that the corresponding classes are probably not available in C:\temp
You can find them here:
https://github.com/danielpalme/ReportGenerator/tree/main/src/Testprojects/CSharp/Project
